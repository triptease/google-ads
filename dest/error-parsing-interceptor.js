"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GaClientError = exports.ExceptionInterceptor = void 0;
const grpc = __importStar(require("@grpc/grpc-js"));
const google_proto_1 = require("../compiled/google-proto");
const FAILURE_KEY = 'google.ads.googleads.v5.errors.googleadsfailure-bin';
const RETRY_STATUS_CODES = [grpc.status.INTERNAL, grpc.status.RESOURCE_EXHAUSTED];
class ExceptionInterceptor {
    constructor() {
        this.requestInterceptor = this.buildRequester();
    }
    intercept(options, nextCall) {
        return new grpc.InterceptingCall(nextCall(options), this.requestInterceptor);
    }
    buildRequester() {
        return (new grpc.RequesterBuilder()
            // tslint:disable-next-line:ban-types
            .withStart((metadata, listener, next) => {
            const newListener = this.buildListener();
            next(metadata, newListener);
        })
            .build());
    }
    buildListener() {
        return (new grpc.ListenerBuilder()
            // tslint:disable-next-line:ban-types
            .withOnReceiveStatus((status, next) => {
            if (status.code !== grpc.status.OK) {
                // TODO: Throw this error instead of returning a new status?
                const error = this.handleGrpcFailure(status);
                const errorStatus = new grpc.StatusBuilder()
                    .withCode(status.code)
                    .withDetails(error.message)
                    .withMetadata(status.metadata)
                    .build();
                next(errorStatus);
            }
            else {
                next(status);
            }
        })
            .build());
    }
    handleGrpcFailure(status) {
        const { code, metadata } = status;
        if (RETRY_STATUS_CODES.includes(code)) {
            /* Throw error if code one of INTERNAL or RESOURCE_EXHAUSTED */
            return new Error(status.details);
        }
        const gaFailures = parseGoogleAdsErrorFromMetadata(metadata);
        if (gaFailures.length === 0) {
            /* Throw error with status details if not a Google Ads API error */
            return new Error(status.details);
        }
        return new GaClientError(gaFailures);
    }
}
exports.ExceptionInterceptor = ExceptionInterceptor;
function parseGoogleAdsErrorFromMetadata(metadata) {
    if (!metadata) {
        return [];
    }
    const failureArray = metadata.get(FAILURE_KEY);
    return failureArray.map(bytes => google_proto_1.google.ads.googleads.v5.errors.GoogleAdsFailure.decode(bytes));
}
class GaClientError extends Error {
    constructor(failures) {
        super(JSON.stringify(failures, null, 2));
        this.failures = failures;
    }
}
exports.GaClientError = GaClientError;
