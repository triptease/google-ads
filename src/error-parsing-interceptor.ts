import grpc from 'grpc';

import { google } from '../compiled/google-proto';

const FAILURE_KEY = 'google.ads.googleads.v5.errors.googleadsfailure-bin';
const RETRY_STATUS_CODES = [grpc.status.INTERNAL, grpc.status.RESOURCE_EXHAUSTED];

type NextCall = (options: grpc.CallOptions) => grpc.InterceptingCall | null;
export type InterceptorMethod = (options: grpc.CallOptions, nextCall: NextCall) => any;

export class ExceptionInterceptor {
  private requestInterceptor: grpc.Requester;

  constructor() {
    this.requestInterceptor = this.buildRequester();
  }

  public intercept(options: grpc.CallOptions, nextCall: NextCall): grpc.InterceptingCall {
    return new grpc.InterceptingCall(nextCall(options), this.requestInterceptor);
  }

  private buildRequester(): grpc.Requester {
    return (
      new grpc.RequesterBuilder()
        // tslint:disable-next-line:ban-types
        .withStart((metadata: grpc.Metadata, listener: grpc.Listener, next: Function) => {
          const newListener = this.buildListener();
          next(metadata, newListener);
        })
        .build()
    );
  }

  private buildListener(): grpc.Listener {
    return (
      new grpc.ListenerBuilder()
        // tslint:disable-next-line:ban-types
        .withOnReceiveStatus((status: grpc.StatusObject, next: Function) => {
          if (status.code !== grpc.status.OK) {
            // TODO: Throw this error instead of returning a new status?
            const error = this.handleGrpcFailure(status);

            const errorStatus = new grpc.StatusBuilder()
              .withCode(status.code)
              .withDetails(error.message)
              .withMetadata(status.metadata)
              .build();

            next(errorStatus);
          } else {
            next(status);
          }
        })
        .build()
    );
  }

  private handleGrpcFailure(status: grpc.StatusObject): Error {
    const { code, metadata } = status;

    if (RETRY_STATUS_CODES.includes(code)) {
      /* Throw error if code one of INTERNAL or RESOURCE_EXHAUSTED */
      return new Error(status.details);
    }

    const gaFailures = parseGoogleAdsErrorFromMetadata(metadata);

    if (gaFailures.length === 0) {
      /* Throw error with status details if not a Google Ads API error */
      return new Error(status.details);
    }

    return new GaClientError(gaFailures);
  }
}

function parseGoogleAdsErrorFromMetadata(
  metadata: grpc.Metadata | undefined
): google.ads.googleads.v5.errors.GoogleAdsFailure[] {
  if (!metadata) {
    return [];
  }

  const failureArray = metadata.get(FAILURE_KEY);

  return failureArray.map(bytes => google.ads.googleads.v5.errors.GoogleAdsFailure.decode(bytes as any));
}

export class GaClientError extends Error {
  constructor(readonly failures: google.ads.googleads.v5.errors.GoogleAdsFailure[]) {
    super(JSON.stringify(failures, null, 2));
  }
}
