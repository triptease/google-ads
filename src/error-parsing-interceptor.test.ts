import fs from 'fs';
import path from 'path';

import { google } from '../compiled/google-proto';
import { GaClientError } from './error-parsing-interceptor';

const GoogleAdsFailure = google.ads.googleads.v2.errors.GoogleAdsFailure;

it('can present google error message from metadata', () => {
  const failureMetadataBytes = fs.readFileSync(path.join(__dirname, 'failure.pb'));
  const obj = GoogleAdsFailure.decode(failureMetadataBytes);

  const err = new GaClientError([obj]);

  expect(err.message).toMatchSnapshot();
});
