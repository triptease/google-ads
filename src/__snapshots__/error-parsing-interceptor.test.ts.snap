// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can present google error message from metadata 1`] = `
"[
  {
    \\"errors\\": [
      {
        \\"errorCode\\": {
          \\"requestError\\": \\"RESOURCE_NAME_MALFORMED\\"
        },
        \\"message\\": \\"Resource name 'invalid' is malformed: expected 'customers/{customer_id}/campaigns/{part_1}'.\\",
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"2\\"
              }
            },
            {
              \\"fieldName\\": \\"ad_group_operation\\"
            },
            {
              \\"fieldName\\": \\"create\\"
            },
            {
              \\"fieldName\\": \\"campaign\\"
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"3\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"4\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"5\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"6\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"7\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"8\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"9\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"10\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"11\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"12\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"13\\"
              }
            }
          ]
        }
      },
      {
        \\"errorCode\\": {
          \\"newResourceCreationError\\": \\"TEMP_ID_RESOURCE_HAD_ERRORS\\"
        },
        \\"message\\": \\"Parent resource with specified temp ID failed validation, so no validation will be done for this child resource.\\",
        \\"trigger\\": {
          \\"int64Value\\": \\"-4\\"
        },
        \\"location\\": {
          \\"fieldPathElements\\": [
            {
              \\"fieldName\\": \\"mutate_operations\\",
              \\"index\\": {
                \\"value\\": \\"14\\"
              }
            }
          ]
        }
      }
    ]
  }
]"
`;
